### ISSUE TITLE
Sidekiq Batch enqueues jobs to development from test environment

### ISSUE THREAD

**Issue by aliismayilov** [mperham/sidekiq#4207]
Ruby version: 2.6.2p47
Sidekiq / Pro / Enterprise version(s): 5.2.5 / 4.0.5 / 1.8.1

Apologies in advance if this question was asked before, as I couldn't find related issue for this. This is a stripped down example:
```ruby
# app/workers/parent_worker.rb
class ParentWorker
  include Sidekiq::Worker

  def perform(parent_id)
    parent = Parent.find(parent_id)
    
    batch = Sidekiq::Batch.new
    batch.on(:success, self.class, 'parent_id' => parent_id)
    batch.jobs do
      parent.children.each do |child|
        ChildWorker.perform_async(child.id)
      end
    end
  end

  def on_success(_status, options)
    # something here
  end
end

# app/workers/child_worker.rb
class ChildWorker
  include Sidekiq::Worker

  def perform(child_id)
    # something here
  end
end
```

After running specs for `ParentWorker` where I call `ParentWorker.new.perform`, I find there are few jobs enqueued in my development environment. This comes from `batch.jobs { ... }`, as it does not happen if I don't enqueue jobs under a batch.

So, I'm curious if this behavior of `Sidekiq::Batch` is a bug or a feature? If it's the latter, what's the recommended approach here?

**Comment by mperham** [comment]
That’s a feature. Batches do not support the Sidekiq::Testing fake/inline modes. I believe that’s documented on the Batches wiki page.

**Comment by aliismayilov** [comment]
I see https://github.com/mperham/sidekiq/wiki/Batches#testing

given that:
1. batch callbacks are not fired during tests;
2. the enqueued jobs under `batch.jobs do ... end` are not run even with `Sidekiq::Testing.inline!`;

why does `Sidekiq::Batch` have to enqueue jobs directly into redis then?

**Comment by mperham** [comment]
Sidekiq::Batch always hits Redis.  Since the callbacks can't be fired, there's no point in adding a testing mode.  If you don't want your dev mode Redis touched, I would configure test to use a different db:

```ruby
config.redis = { db: 14 } if Rails.env.test?
```

**Comment by diaclavijo** [comment]
Your documentation https://github.com/mperham/sidekiq/wiki/Testing#testing-batches states:

![image](https://user-images.githubusercontent.com/5384036/87555462-c681d880-c6b5-11ea-996d-75813f817b97.png)

When I read that what I expected is that the callback done would not be fired under Inline mode but the  jobs will be executed. 

Even when I read the other docs I keep understanding the same. https://github.com/mperham/sidekiq/wiki/Batches#testing

It is clear for me that the callbacks will be not executed but I understood when I read it that the jobs would be.

**Comment by victormier** [comment]
@mperham I agree with @diaclavijo diaclavijo this is unclear. Jobs in batches seem to not be running even using inline mode. Documentation example seems to not be working.

What would be the right way to test that the jobs are run and executed?

```ruby
def do_something
  batch = Sidekiq::Batch.new
  batch.jobs do
    SimpleWorker.perform_async("foo")
    SimpleWorker.perform_async("bar")
  end
end
```

**Comment by ouranos** [comment]
I thought I'd comment here as I just encountered a really confusing behavior. 
Unlike what was stated here, jobs in batch were enqueued in rspec (without redis).
But with cucumber I had the same behavior as stated above:

> Sidekiq::Batch always hits Redis.

It turns out that the `rspec-sidekiq` gem enables [testing batches without redis](https://github.com/philostler/rspec-sidekiq#batches) 🤦 

So if you want to test that the jobs in a batch are queued/performed, you can use this gem (but you still have to manually trigger callbacks). This will respect the testing strategy (fake/inline).
